ingestor/shared/src/main/scala/za/co/absa/hyperdrive/shared/test/utils/SparkTestUtils.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ *  Copyright 2019 ABSA Group Limited
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.shared.test.utils
+
+import org.apache.spark.sql.types.{ArrayType, StructField, StructType}
+
+object SparkTestUtils {
+
+  def areAllFieldsNullable(schema: StructType): Boolean = {
+    schema.fields.forall(isNullable(_))
+  }
+
+  private def isNullable(field: StructField): Boolean = {
+    val inspectionResult = field match {
+      case dt: StructType => areAllFieldsNullable(dt)
+      case dt: ArrayType => areAllFieldsNullable(dt)
```

Comment:
Hmm... I think you should have used `field.dataType` here? There are the following compiler warnings
```
warning: fruitless type test: a value of type org.apache.spark.sql.types.StructField cannot also be a org.apache.spark.sql.types.StructType
warning: fruitless type test: a value of type org.apache.spark.sql.types.StructField cannot also be a org.apache.spark.sql.types.ArrayType
```
But anyway, imo the logic is non-trivial and maybe requires a test itself. Wouldn't it be easier to simply explicitly check every field in `TestTestAllNullableParquetStreamWriter` and `TestSparkUtils`? E.g.
`assert(dummyStreamWriter.capturedDataframe.schema.fields(0).nullable)`

Reply:
Yes, this is why I had closed the PR, there was a missing suite for the `SparkTestUtils` itself.

The problem with this logic is the presence of inner fields and the types they can assume. There are `StructFields`, `StructTypes` and `ArrayTypes`. The latter two can have any combinations of all three so we need to go recursively, and also check this behaviour.

We have now two utils, `SparkUtils` which provide methods to change nullabilities, and `SparkTestUtils` which assist with the checking of schemas nullabilities.